### PR TITLE
[SPARK-27612][PYTHON] Use Python's default protocol instead of highest protocol

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR partially reverts https://github.com/apache/spark/pull/20691

After we changed the Python protocol to highest ones, seems like it introduced a correctness bug. This potentially affects all Python related code paths.

I suspect a bug related to Pryolite (maybe opcodes `MEMOIZE`, `FRAME` and/or our `RowPickler`). I would like to stick to default protocol for now and investigate the issue separately.

I will separately investigate later to bring highest protocol back.

Update:
I can almost confirm that this is a bug in Pyrolite.

1. Python lists are chunked via `BatchedSerializer` at `parallelize` (in `createDataFrame`), for instance, as below (each line is each batch):

    ```bash
    [[1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4]]
    [[1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4]]
    ...
    [[1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4]]
    ```

    Here is opcodes for each batch (separated by `===..`) with protocol 4:
    <details>

    ```
    ==============================
        0: \x80 PROTO      4
        2: \x95 FRAME      47
       11: ]    EMPTY_LIST
       12: \x94 MEMOIZE    (as 0)
       13: (    MARK
       14: ]        EMPTY_LIST
       15: \x94     MEMOIZE    (as 1)
       16: (        MARK
       17: K            BININT1    1
       19: K            BININT1    2
       21: K            BININT1    3
       23: K            BININT1    4
       25: e            APPENDS    (MARK at 16)
       26: \x85     TUPLE1
       27: \x94     MEMOIZE    (as 2)
       28: h        BINGET     1
       30: \x85     TUPLE1
       31: \x94     MEMOIZE    (as 3)
       32: h        BINGET     1
       34: \x85     TUPLE1
       35: \x94     MEMOIZE    (as 4)
       36: h        BINGET     1
       38: \x85     TUPLE1
       39: \x94     MEMOIZE    (as 5)
       40: h        BINGET     1
       42: \x85     TUPLE1
       43: \x94     MEMOIZE    (as 6)
       44: h        BINGET     1
       46: \x85     TUPLE1
       47: \x94     MEMOIZE    (as 7)
       48: h        BINGET     1
       50: \x85     TUPLE1
       51: \x94     MEMOIZE    (as 8)
       52: h        BINGET     1
       54: \x85     TUPLE1
       55: \x94     MEMOIZE    (as 9)
       56: e        APPENDS    (MARK at 13)
       57: .    STOP
    highest protocol among opcodes = 4
    ==============================
    ==============================
        0: \x80 PROTO      4
        2: \x95 FRAME      47
       11: ]    EMPTY_LIST
       12: \x94 MEMOIZE    (as 0)
       13: (    MARK
       14: ]        EMPTY_LIST
       15: \x94     MEMOIZE    (as 1)
       16: (        MARK
       17: K            BININT1    1
       19: K            BININT1    2
       21: K            BININT1    3
       23: K            BININT1    4
       25: e            APPENDS    (MARK at 16)
       26: \x85     TUPLE1
       27: \x94     MEMOIZE    (as 2)
       28: h        BINGET     1
       30: \x85     TUPLE1
       31: \x94     MEMOIZE    (as 3)
       32: h        BINGET     1
       34: \x85     TUPLE1
       35: \x94     MEMOIZE    (as 4)
       36: h        BINGET     1
       38: \x85     TUPLE1
       39: \x94     MEMOIZE    (as 5)
       40: h        BINGET     1
       42: \x85     TUPLE1
       43: \x94     MEMOIZE    (as 6)
       44: h        BINGET     1
       46: \x85     TUPLE1
       47: \x94     MEMOIZE    (as 7)
       48: h        BINGET     1
       50: \x85     TUPLE1
       51: \x94     MEMOIZE    (as 8)
       52: h        BINGET     1
       54: \x85     TUPLE1
       55: \x94     MEMOIZE    (as 9)
       56: e        APPENDS    (MARK at 13)
       57: .    STOP
    highest protocol among opcodes = 4
    ==============================
    ...
    ==============================
        0: \x80 PROTO      4
        2: \x95 FRAME      31
       11: ]    EMPTY_LIST
       12: \x94 MEMOIZE    (as 0)
       13: (    MARK
       14: ]        EMPTY_LIST
       15: \x94     MEMOIZE    (as 1)
       16: (        MARK
       17: K            BININT1    1
       19: K            BININT1    2
       21: K            BININT1    3
       23: K            BININT1    4
       25: e            APPENDS    (MARK at 16)
       26: \x85     TUPLE1
       27: \x94     MEMOIZE    (as 2)
       28: h        BINGET     1
       30: \x85     TUPLE1
       31: \x94     MEMOIZE    (as 3)
       32: h        BINGET     1
       34: \x85     TUPLE1
       35: \x94     MEMOIZE    (as 4)
       36: h        BINGET     1
       38: \x85     TUPLE1
       39: \x94     MEMOIZE    (as 5)
       40: e        APPENDS    (MARK at 13)
       41: .    STOP
    highest protocol among opcodes = 4
    ==============================
    ```
    </details>

2. Those batches become binary in an RDD at `parallelize` - so far results look fine

3. `rdd._to_java_object_rdd` -> `SerDeUtil.pythonToJava` -> `Unpickler.loads`

4. Here, Pryolite `Unpickler.loads` converts each Python object batch into Java objects

5. The converted Java objects look a bit odd from Pryolite as below (see the last three lists):

    ```
    [1, 2, 3, 4]
    [1, 2, 3, 4]
    ...
    [[[1, 2, 3, 4]], [[1, 2, 3, 4]]]
    [[[1, 2, 3, 4]], [[1, 2, 3, 4]]]
    [[[1, 2, 3, 4]], [[1, 2, 3, 4]]]
    ```

6. Last three nested lists are ignored via `EvaluatePython.makeFromJava` and becomes `null` -> `None`.

7. So the Jenkins test was failed at https://github.com/apache/spark/pull/24519#issuecomment-488913729 as below:

    ```
    Traceback (most recent call last):
      File "/home/jenkins/workspace/SparkPullRequestBuilder/python/pyspark/sql/tests/test_serde.py", line 134, in test_int_array_serialization
        self.assertEqual(len(list(filter(lambda r: None in r.value, df.collect()))), 0)
    AssertionError: 3 != 0
    ```

## How was this patch tested?

Unittest was added.

```bash
./run-tests --python-executables=python3.7 --testname "pyspark.sql.tests.test_serde SerdeTests.test_int_array_serialization"
```

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @BryanCutler and @viirya 

This PR could not identify the root cause yet (but only my rough hypothesis). I would like to get this in first and then do the investigation separately.

One concern about using default protocol is that it's not tested in cloudpickle. I opened a PR to add a env to control that and test it (https://github.com/cloudpipe/cloudpickle/pull/265).

File: python/pyspark/serializers.py [hunk 1]
```diff
@@ -62,11 +62,12 @@
 if sys.version < '3':
     import cPickle as pickle
     from itertools import izip as zip, imap as map
+    pickle_protocol = 2
 else:
     import pickle
     basestring = unicode = str
     xrange = range
-pickle_protocol = pickle.HIGHEST_PROTOCOL
+    pickle_protocol = 3
```
Comment (HyukjinKwon) [inline python/pyspark/serializers.py]: We can use `pickle.DEFAULT_PROTOCOL` too but let me stick with constant since seems protocol 4 has this bug.

Comment (dongjoon-hyun) [reply python/pyspark/serializers.py]: It looks nice and solid.

**Review by viirya — COMMENTED** [review]
Agreed that this can be a temporary fix for the correctness issue. We can investigate it more for root cause after this.

**Comment by viirya** [comment]
The test failed? Can't default protocol fix this?

**Comment by HyukjinKwon** [comment]
Oh, https://github.com/apache/spark/pull/24519#issuecomment-488913729 was intended to be failed. I wanted to make sure it's failed in Jenkins too.

**Comment by dongjoon-hyun** [comment]
The last commit is still under testing~

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by viirya** [comment]
Oh, don't see that. Thanks.

**Review by viirya — APPROVED** [review]

File: python/pyspark/serializers.py (hunk 1 above)
Comment (HyukjinKwon) [reply python/pyspark/serializers.py]: Do you mean the bug issue related to the root cause somewhere? (I think) it's more like an issue within Pryolite library .. I am not 100% sure yet. I will update that when I have it. I am looking into this to identify the cause.

**Comment by HyukjinKwon** [comment]
I can almost confirm that this is a bug in Pyrolite.

1. Python lists are chunked via `BatchedSerializer` at `parallelize`, for instance, as below (each line is each batch):

    ```bash
    [[1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4]]
    [[1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4]]
    ...
    [[1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4]]
    ```

    Here is opcodes for the batches with protocol 4:
    <details>

    ```
    ==============================
        0: \x80 PROTO      4
        2: \x95 FRAME      47
       11: ]    EMPTY_LIST
       12: \x94 MEMOIZE    (as 0)
       13: (    MARK
       14: ]        EMPTY_LIST
       15: \x94     MEMOIZE    (as 1)
       16: (        MARK
       17: K            BININT1    1
       19: K            BININT1    2
       21: K            BININT1    3
       23: K            BININT1    4
       25: e            APPENDS    (MARK at 16)
       26: \x85     TUPLE1
       27: \x94     MEMOIZE    (as 2)
       28: h        BINGET     1
       30: \x85     TUPLE1
       31: \x94     MEMOIZE    (as 3)
       32: h        BINGET     1
       34: \x85     TUPLE1
       35: \x94     MEMOIZE    (as 4)
       36: h        BINGET     1
       38: \x85     TUPLE1
       39: \x94     MEMOIZE    (as 5)
       40: h        BINGET     1
       42: \x85     TUPLE1
       43: \x94     MEMOIZE    (as 6)
       44: h        BINGET     1
       46: \x85     TUPLE1
       47: \x94     MEMOIZE    (as 7)
       48: h        BINGET     1
       50: \x85     TUPLE1
       51: \x94     MEMOIZE    (as 8)
       52: h        BINGET     1
       54: \x85     TUPLE1
       55: \x94     MEMOIZE    (as 9)
       56: e        APPENDS    (MARK at 13)
       57: .    STOP
    highest protocol among opcodes = 4
    ==============================
    ==============================
        0: \x80 PROTO      4
        2: \x95 FRAME      47
       11: ]    EMPTY_LIST
       12: \x94 MEMOIZE    (as 0)
       13: (    MARK
       14: ]        EMPTY_LIST
       15: \x94     MEMOIZE    (as 1)
       16: (        MARK
       17: K            BININT1    1
       19: K            BININT1    2
       21: K            BININT1    3
       23: K            BININT1    4
       25: e            APPENDS    (MARK at 16)
       26: \x85     TUPLE1
       27: \x94     MEMOIZE    (as 2)
       28: h        BINGET     1
       30: \x85     TUPLE1
       31: \x94     MEMOIZE    (as 3)
       32: h        BINGET     1
       34: \x85     TUPLE1
       35: \x94     MEMOIZE    (as 4)
       36: h        BINGET     1
       38: \x85     TUPLE1
       39: \x94     MEMOIZE    (as 5)
       40: h        BINGET     1
       42: \x85     TUPLE1
       43: \x94     MEMOIZE    (as 6)
       44: h        BINGET     1
       46: \x85     TUPLE1
       47: \x94     MEMOIZE    (as 7)
       48: h        BINGET     1
       50: \x85     TUPLE1
       51: \x94     MEMOIZE    (as 8)
       52: h        BINGET     1
       54: \x85     TUPLE1
       55: \x94     MEMOIZE    (as 9)
       56: e        APPENDS    (MARK at 13)
       57: .    STOP
    highest protocol among opcodes = 4
    ==============================
    ...
    ==============================
        0: \x80 PROTO      4
        2: \x95 FRAME      31
       11: ]    EMPTY_LIST
       12: \x94 MEMOIZE    (as 0)
       13: (    MARK
       14: ]        EMPTY_LIST
       15: \x94     MEMOIZE    (as 1)
       16: (        MARK
       17: K            BININT1    1
       19: K            BININT1    2
       21: K            BININT1    3
       23: K            BININT1    4
       25: e            APPENDS    (MARK at 16)
       26: \x85     TUPLE1
       27: \x94     MEMOIZE    (as 2)
       28: h        BINGET     1
       30: \x85     TUPLE1
       31: \x94     MEMOIZE    (as 3)
       32: h        BINGET     1
       34: \x85     TUPLE1
       35: \x94     MEMOIZE    (as 4)
       36: h        BINGET     1
       38: \x85     TUPLE1
       39: \x94     MEMOIZE    (as 5)
       40: e        APPENDS    (MARK at 13)
       41: .    STOP
    highest protocol among opcodes = 4
    ==============================
    ```
    </details>

2. Those batches become binary in an RDD at `parallelize` - so far results look fine

3. `rdd._to_java_object_rdd` -> `SerDeUtil.pythonToJava` -> `Unpickler.loads`

4. Here, Pryolite `Unpickler.loads` converts each Python object batch into Java objects

5. The converted Java objects look a bit odd from Pryolite as below:

    ```
    [1, 2, 3, 4]
    [1, 2, 3, 4]
    ...
    [[[1, 2, 3, 4]], [[1, 2, 3, 4]]]
    [[[1, 2, 3, 4]], [[1, 2, 3, 4]]]
    [[[1, 2, 3, 4]], [[1, 2, 3, 4]]]
    ```

6. Nested last three list are ignored via `EvaluatePython.makeFromJava` and becomes `null` -> `None`.

**Comment by HyukjinKwon** [comment]
Not sure if there's quick fix because notable opcodes that appeared from both protocol 4 and current reproducer is only `MEMORIZE`. We should peek `memo` at `Unpickler` which is private in Pryolite 4.13. It's protected in 4.17 so maybe we can later upgrade Pryolite and debug further.

**Comment by HyukjinKwon** [comment]
Merged to master.

Thanks for review, @dongjoon-hyun and @viirya

**Comment by dongjoon-hyun** [comment]
Hi, @HyukjinKwon .
This issue is marked as a correctness issue. Do we need to backport this?

**Pull request closed without merging**